File: src/main/java/io/pivotal/pal/tracker/TimeEntryController.java
package io.pivotal.pal.tracker;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/time-entries")
public class TimeEntryController {
  private TimeEntryRepository repository;

  public TimeEntryController(TimeEntryRepository repository) {
    this.repository = repository;
  }

  @PostMapping
  public ResponseEntity<TimeEntry> create(@RequestBody TimeEntry timeEntryToCreate) {
    return new ResponseEntity(repository.create(timeEntryToCreate), HttpStatus.CREATED);
  }

  @GetMapping("/{id}")
  public ResponseEntity<TimeEntry> read(@PathVariable("id") long id) {
    TimeEntry found = repository.find(id);
    return (found == null)
        ? new ResponseEntity(HttpStatus.NOT_FOUND)
        : ResponseEntity.ok(found);
  }

  @GetMapping
  public ResponseEntity<List<TimeEntry>> list() {
    return ResponseEntity.ok(repository.list());
  }

  @PutMapping("/{id}")
  public ResponseEntity update(@PathVariable("id") long id, @RequestBody TimeEntry timeEntry) {
    TimeEntry updated = repository.update(id, timeEntry);
    return updated == null
        ? new ResponseEntity(HttpStatus.NOT_FOUND)
        : ResponseEntity.ok(updated);
  }

  @DeleteMapping("/{id}")
  public ResponseEntity delete(@PathVariable("id") long id) {
    repository.delete(id);
    return new ResponseEntity(HttpStatus.NO_CONTENT);
  }
}

File: src/main/kotlin/io/pivotal/pal/tracker/TimeEntryController.kt
package io.pivotal.pal.tracker

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/time-entries")
class TimeEntryController(private val repository: TimeEntryRepository){

    @PostMapping
    fun create(@RequestBody expected: TimeEntry): ResponseEntity<TimeEntry> {
        val entry = repository.create(expected)
        return ResponseEntity(entry, HttpStatus.CREATED)
    }

    @GetMapping("/{id}")
    fun read(@PathVariable id: Long): ResponseEntity<TimeEntry?> {
        val found = repository.find(id)
        return when (found != null) {
            true -> ResponseEntity(found, HttpStatus.OK)
            false -> ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }

    @GetMapping
    fun list(): ResponseEntity<List<TimeEntry>> {
        return ResponseEntity(repository.list(), HttpStatus.OK)
    }

    @PutMapping("/{id}")
    fun update(@PathVariable("id") id:Long, @RequestBody expected:TimeEntry): ResponseEntity<TimeEntry?> {
        val updated= repository.update(id, expected)
        return when(updated != null) {
            true -> ResponseEntity(updated, HttpStatus.OK)
            false -> ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable("id") id:Long): ResponseEntity<Unit> {
        val deleted = repository.delete(id)
        return when(deleted != null) {
            true -> ResponseEntity(HttpStatus.OK)
            false -> ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }

}